src/black/__init__.py

```diff
@@ -659,25 +664,27 @@ def get_sources(
             sources.add(p)
         elif p.is_dir():
             if exclude is None:
-                exclude = re_compile_maybe_verbose(DEFAULT_EXCLUDES)
-                gitignore = get_gitignore(root)
+                if _exclude is None:
+                    _exclude = re_compile_maybe_verbose(DEFAULT_EXCLUDES)
+                if gitignore is None:
+                    gitignore = get_gitignore(root)
                 p_gitignore = get_gitignore(p)
                 # No need to use p's gitignore if it is identical to root's gitignore
                 # (i.e. root and p point to the same directory).
-                if gitignore != p_gitignore:
-                    gitignore += p_gitignore
-            else:
-                gitignore = None
+                if gitignore == p_gitignore:
+                    _gitignore = gitignore
+                else:
+                    _gitignore = gitignore + p_gitignore
             sources.update(
                 gen_python_files(
                     p.iterdir(),
                     ctx.obj["root"],
                     include,
-                    exclude,
+                    cast(Pattern[str], _exclude),
```

Comment:
Instead of casting, we should assert that _exclude is not None. cast is unsafe.

Reply:
I'm not sure how to assert that. The value will never be `None`, but is a limitation (?) of mypy https://github.com/psf/black/pull/3336/files#r997119695

But, this issue should go away if we just initialize `_exclude` as `re_compile_maybe_verbose(DEFAULT_EXCLUDES)`, as you suggested. Then, casting nor asserting should be necessary (mypy will correctly identify `_exclude` as `Pattern[str]`)